AppDB/appscale/datastore/cassandra_env/large_batch.py

```diff
@@ -214,3 +224,215 @@ def claim(self):
+    self._tornado_cassandra = TornadoCassandra(self._session)
+    self._prepared_statements = {}
+
+  @gen.coroutine
+  def resolve(self, txid, composite_indexes):
+    """ Resolves a large batch for a given transaction.
+
+    Args:
+      txid: An integer specifying a transaction ID.
+      composite_indexes: A list of CompositeIndex objects.
+    """
+    txid_hash = tx_partition(self.project_id, txid)
+    new_op_id = uuid.uuid4()
+    try:
+      batch_status = yield self._get_status(txid_hash)
+    except BatchNotFound:
+      # Make sure another process doesn't try to commit the transaction.
+      yield self._insert(txid_hash, new_op_id)
+      raise gen.Return()
+
+    old_op_id = batch_status.op_id
+    yield self._update_op_id(txid_hash, batch_status.applied, old_op_id,
+                             new_op_id)
+
+    if batch_status.applied:
+      # Make sure all the mutations in the batch have been applied.
+      yield self._apply_mutations(txid, composite_indexes)
+
+  @gen.coroutine
+  def cleanup(self, txid):
+    """ Cleans up the metadata from the finished batch.
+
+    Args:
+      txid: An integer specifying a transaction ID.
+    """
+    txid_hash = tx_partition(self.project_id, txid)
+    yield self._delete_mutations(txid)
+    yield self._delete_status(txid_hash)
+
+  def _get_prepared(self, statement):
+    """ Caches prepared statements.
+
+    Args:
+      statement: A string containing a Cassandra statement.
+    """
+    if statement not in self._prepared_statements:
+      self._prepared_statements[statement] = self._session.prepare(statement)
+
+    return self._prepared_statements[statement]
+
+  @gen.coroutine
+  def _get_status(self, txid_hash):
+    """ Gets the current status of a large batch.
+
+    Args:
+      txid_hash: A byte array identifying the transaction.
+    Returns:
+      A Cassandra result for the batch entry.
+    """
+    statement = self._get_prepared("""
+      SELECT applied, op_id FROM batch_status
+      WHERE txid_hash = ?
+    """)
+    bound_statement = statement.bind((txid_hash,))
+    bound_statement.consistency_level = ConsistencyLevel.SERIAL
+    bound_statement.retry_policy = BASIC_RETRIES
+    results = yield self._tornado_cassandra.execute(bound_statement)
+    try:
+      raise gen.Return(results[0])
+    except IndexError:
+      raise BatchNotFound('Batch not found')
+
+  @gen.coroutine
+  def _insert(self, txid_hash, op_id):
+    """ Claims the large batch.
+
+    Args:
+      txid_hash: A byte array identifying the transaction.
+      op_id: A uuid4 specifying the process ID.
+    """
+    statement = self._get_prepared("""
+      INSERT INTO batch_status (txid_hash, applied, op_id)
+      VALUES (?, ?, ?)
+      IF NOT EXISTS
+    """)
+    bound_statement = statement.bind((txid_hash, False, op_id))
+    bound_statement.retry_policy = NO_RETRIES
+    results = yield self._tornado_cassandra.execute(bound_statement)
+    if not results[0].applied:
```

Comment:
A lightweight transaction insert should always have exactly one row in the result set that indicates whether or not the statement was applied. Typically, it's clearer to use the `.was_applied` method instead of `[0].applied`, but I made the mistake of naming one of the columns "applied", and that prevents the method from working as expected.

The execute method is generic, and can be used with any type of statement. If a select statement returned no rows, we would not need to raise an exception.

Reply:
Is there any other possible reason for the result to be not applied? Or BatchNowOwned is one possible cause?

Reply:
The insert can fail due to a connection issue (in which case, the above yield will raise an exception), but if we get a response and the statement was not applied, it can only be because the row already exists (the if clause of the statement). If the row already exists, a different process must have inserted it.